File: myExpenses/src/main/java/org/totschnig/myexpenses/util/PermissionHelper.java
package org.totschnig.myexpenses.util;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Binder;
import android.os.Build;

import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.preference.PrefKey;

import java.io.File;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import static android.content.pm.PackageManager.PERMISSION_DENIED;
import static android.content.pm.PackageManager.PERMISSION_GRANTED;

public class PermissionHelper {
  public static final int PERMISSIONS_REQUEST_WRITE_CALENDAR = 1;
  public static final int PERMISSIONS_REQUEST_STORAGE = 2;

  private PermissionHelper() {}

  public enum PermissionGroup {
    STORAGE(externalReadPermissionCompat(), PrefKey.STORAGE_PERMISSION_REQUESTED, PERMISSIONS_REQUEST_STORAGE),
    CALENDAR(new String[]{Manifest.permission.WRITE_CALENDAR, Manifest.permission.READ_CALENDAR}, PrefKey.CALENDAR_PERMISSION_REQUESTED, PERMISSIONS_REQUEST_WRITE_CALENDAR);

    public final String[] androidPermissions;
    public final PrefKey prefKey;
    public final int requestCode;

    PermissionGroup(String[] androidPermission, PrefKey prefKey, int requestCode) {
      this.androidPermissions = androidPermission;
      this.prefKey = prefKey;
      this.requestCode = requestCode;
    }

    public static PermissionGroup fromRequestCode(int requestCode) {
      if (requestCode == STORAGE.requestCode) return STORAGE;
      if (requestCode == CALENDAR.requestCode) return CALENDAR;
      throw new IllegalArgumentException("Undefined requestCode " + requestCode);
    }

    /**
     *
     * @param context
     * @return true if all of our {@link #androidPermissions} are granted
     */
    public boolean hasPermission(Context context) {
      for (String permission: androidPermissions) {
        if (!PermissionHelper.hasPermission(context, permission)) {
          return false;
        }
      }
      return true;
    }

    /**
     *
     * @param context
     * @return true if {@link ActivityCompat#shouldShowRequestPermissionRationale(Activity, String)}
     * returns true for any of our {@link #androidPermissions}
     */
    public boolean shouldShowRequestPermissionRationale(Activity context) {
      for (String permission: androidPermissions) {
        if (ActivityCompat.shouldShowRequestPermissionRationale(context, permission)) {
          return true;
        }
      }
      return false;
    }
  }

  public static boolean hasPermission(Context context, String permission) {
    return getSelfPermissionSafe(context, permission) == PERMISSION_GRANTED;
  }

  public static boolean hasCalendarPermission(Context context) {
    return PermissionGroup.CALENDAR.hasPermission(context);
  }

  public static boolean hasExternalReadPermission(Context context) {
    return PermissionGroup.STORAGE.hasPermission(context);
  }

  /**
   * Wrapper around {@link ContextCompat#checkSelfPermission(Context, String)}
   * Workaround for RuntimeException
   * See https://github.com/permissions-dispatcher/PermissionsDispatcher/pull/108/files
   * @param context context
   * @param permission permission
   * @return returns true if context has access to the given permission, false otherwise.
   */
  public static int getSelfPermissionSafe(Context context, String permission) {
    try {
      return ContextCompat.checkSelfPermission(context, permission);
    } catch (RuntimeException t) {
      return PERMISSION_DENIED;
    }
  }

  public static String[] externalReadPermissionCompat() {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
      return new String[] {Manifest.permission.READ_EXTERNAL_STORAGE};
    } else {
      return new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE};
    }
  }

  public static boolean canReadUri(Uri uri, Context context) {
    if ("file".equals(uri.getScheme())) {
      File file = new File(uri.getPath());
      return file.exists() && file.canRead();
    }
    return AppDirHelper.getFileProviderAuthority().equals(uri.getAuthority()) ||
        hasExternalReadPermission(context) ||
        context.checkUriPermission(uri, Binder.getCallingPid(), Binder.getCallingUid(),
            Intent.FLAG_GRANT_READ_URI_PERMISSION) == PERMISSION_GRANTED;
  }

  public static PrefKey permissionRequestedKey(int requestCode) {
    return PermissionGroup.fromRequestCode(requestCode).prefKey;
  }

  public static CharSequence permissionRequestRationale(Context context, int requestCode) {
    switch (requestCode) {
      case PERMISSIONS_REQUEST_WRITE_CALENDAR:
        return Utils.getTextWithAppName(context, R.string.calendar_permission_required);
      case PERMISSIONS_REQUEST_STORAGE:
        return context.getString(R.string.storage_permission_required);
    }
    throw new IllegalArgumentException("Undefined requestCode " + requestCode);
  }

  public static boolean allGranted(int[] grantResults) {
    if (grantResults.length == 0) {
      return false;
    }
    for (int result: grantResults) {
      if (result != PackageManager.PERMISSION_GRANTED) {
        return false;
      }
    }
    return true;
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/util/PermissionHelper.kt
package org.totschnig.myexpenses.util

import android.Manifest
import android.annotation.TargetApi
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Binder
import android.os.Build
import androidx.annotation.IntDef
import androidx.annotation.RequiresApi
import com.vmadalin.easypermissions.EasyPermissions
import org.totschnig.myexpenses.R
import java.io.File


object PermissionHelper {
    @IntDef(PERMISSIONS_REQUEST_WRITE_CALENDAR, PERMISSIONS_REQUEST_NOTIFICATIONS_WEBUI,
        PERMISSIONS_REQUEST_NOTIFICATIONS_SYNC, PERMISSIONS_REQUEST_NOTIFICATIONS_PLANNER,
        PERMISSIONS_REQUEST_NOTIFICATIONS_AUTO_BACKUP, PERMISSIONS_REQUEST_STORAGE
    )
    @Retention(AnnotationRetention.SOURCE)
    annotation class PermissionRequestCode

    const val PERMISSIONS_REQUEST_WRITE_CALENDAR = 1
    const val PERMISSIONS_REQUEST_NOTIFICATIONS_WEBUI = 2
    const val PERMISSIONS_REQUEST_NOTIFICATIONS_SYNC = 3
    const val PERMISSIONS_REQUEST_NOTIFICATIONS_PLANNER = 4
    const val PERMISSIONS_REQUEST_NOTIFICATIONS_AUTO_BACKUP = 5
    const val PERMISSIONS_REQUEST_STORAGE = 6

    @JvmStatic
    fun hasCalendarPermission(context: Context) = PermissionGroup.CALENDAR.hasPermission(context)

    @JvmStatic
    fun canReadUri(uri: Uri, context: Context): Boolean {
        if ("file" == uri.scheme) {
            uri.path?.let {
                return with(File(it)) {
                    exists() && canRead()
                }
            }
            return false
        }
        return AppDirHelper.getFileProviderAuthority(context) == uri.authority || context.checkUriPermission(
            uri, Binder.getCallingPid(), Binder.getCallingUid(),
            Intent.FLAG_GRANT_READ_URI_PERMISSION
        ) == PackageManager.PERMISSION_GRANTED
    }

    @TargetApi(Build.VERSION_CODES.TIRAMISU)
    fun getRationale(context: Context, @PermissionRequestCode requestCode: Int, vararg group: PermissionGroup): String =
        when (requestCode) {
            PERMISSIONS_REQUEST_WRITE_CALENDAR -> {
                group.joinToString(" ") {
                    Utils.getTextWithAppName(
                        context, when(it) {
                            PermissionGroup.CALENDAR -> R.string.calendar_permission_required
                            PermissionGroup.NOTIFICATION -> R.string.notifications_permission_required_planner
                            else -> throw IllegalArgumentException()
                        }).toString()
                }
            }
            PERMISSIONS_REQUEST_NOTIFICATIONS_WEBUI -> context.getString(R.string.notifications_permission_required_webui)
            PERMISSIONS_REQUEST_NOTIFICATIONS_SYNC -> context.getString(R.string.notifications_permission_required_sync)
            PERMISSIONS_REQUEST_NOTIFICATIONS_PLANNER -> Utils.getTextWithAppName(context, R.string.notifications_permission_required_planner).toString()
            PERMISSIONS_REQUEST_NOTIFICATIONS_AUTO_BACKUP -> context.getString(R.string.notifications_permission_required_auto_backup)
            PERMISSIONS_REQUEST_STORAGE -> "In order to save images to external storage you must grant Storage Permission"
            else -> throw IllegalArgumentException()
        }

    enum class PermissionGroup(
        val androidPermissions: List<String>
    ) {
        STORAGE(
            listOf(Manifest.permission.WRITE_EXTERNAL_STORAGE)
        ),
        CALENDAR(
            listOf(Manifest.permission.WRITE_CALENDAR, Manifest.permission.READ_CALENDAR)
        ),
        @RequiresApi(Build.VERSION_CODES.TIRAMISU)
        NOTIFICATION(
            listOf(Manifest.permission.POST_NOTIFICATIONS)
        )
        ;

        /**
         * @return true if all of our [.androidPermissions] are granted
         */
        fun hasPermission(context: Context): Boolean {
            return EasyPermissions.hasPermissions(context, *androidPermissions.toTypedArray())
        }

        companion object {
            fun fromPermission(permission: String): PermissionGroup {
                return PermissionGroup.entries.first { it.androidPermissions.contains(permission) }
            }
        }
    }
}